ostd/src/mm/dma/dma_stream.rs

```diff
@@ -1,212 +1,330 @@
-            DmaType::Direct => {
-                #[cfg(target_arch = "x86_64")]
-                crate::arch::if_tdx_enabled!({
-                    // SAFETY:
-                    //  - The address of a `USegment` is always page aligned.
-                    //  - A `USegment` always points to normal physical memory, so the address
-                    //    range falls in the GPA limit.
-                    //  - A `USegment` always points to normal physical memory, so all the pages
-                    //    are contained in the linear mapping.
-                    //  - The pages belong to a `USegment`, so they're all untyped memory.
-                    unsafe {
-                        crate::arch::tdx_guest::unprotect_gpa_range(paddr, frame_count).unwrap();
-                    }
-                });
-                paddr as Daddr
-            }
-            DmaType::Iommu => {
-                for i in 0..frame_count {
-                    let paddr = paddr + (i * PAGE_SIZE);
-                    // SAFETY: the `paddr` is restricted by the `paddr` and `frame_count` of the `segment`.
-                    unsafe {
-                        iommu::map(paddr as Daddr, paddr).unwrap();
-                    }
-                }
-                paddr as Daddr
-            }
+        let (inner, paddr) = if (can_sync_dma() || is_cache_coherent) && !has_tdx {
+            let paddr = segment.paddr();
+
+            (Inner::Segment(segment), paddr)
+        } else {
+            let (kva, paddr) = alloc_kva(size, has_tdx, is_cache_coherent)?;
+
+            (Inner::Both(kva, paddr, segment), paddr)
         };
 
+        let paddr_range = paddr..paddr + size;
+
+        // SAFETY: The physical address range is untyped DMA memory before `drop`.
+        #[cfg(any(debug_assertions, all(target_arch = "x86_64", feature = "cvm_guest")))]
+        unsafe {
+            alloc_unprotect_physical_range(&paddr_range)
+        };
+
+        // SAFETY: The physical address range is untyped DMA memory before `drop`.
+        let map_daddr = unsafe { dma_remap(&paddr_range) };
+
         Ok(Self {
-            segment,
-            start_daddr,
+            inner,
+            map_daddr,
             is_cache_coherent,
-            direction,
+            _phantom: PhantomData,
         })
     }
 
-    /// Gets the underlying [`USegment`].
+    /// Synchronizes the streaming DMA mapping data from the device.
     ///
-    /// Usually, the CPU side should not access the memory
-    /// after the DMA mapping is established because
-    /// there is a chance that the device is updating
-    /// the memory. Do this at your own risk.
-    pub fn segment(&self) -> &USegment {
-        &self.segment
-    }
+    /// This method should be called when the data of the streaming DMA mapping
+    /// has been updated by the device side. And the CPU side needs to call the
+    /// [`Self::sync_from_device`] method before reading data (e.g., using
+    /// [`read_bytes`]).
```

Comment:
```suggestion
    /// has been updated by the device side. Before the CPU side starts to read
    /// (e.g., using [`read_bytes`]), it must call the [`Self::sync_from_device`]
    /// method first.
```

This is slightly more consistent to the `sync_to_device` docs:
```rust
    /// has been updated by the CPU side (e.g., using [`write_bytes`]). Before
    /// the CPU side notifies the device side to read, it must call the
    /// [`Self::sync_to_device`] method first.
```